File: yr2/Semester1/Programming Languagues (F28PL)/CW2/CW2Q4.py
# Q4a

# s1 and s2 represent tuple
def threatening(s1, s2):
	# checks Rook moves
	if s1[0] == s2[0] or s1[1] == s2[1]:
		return True


	# list of knight moves
	knight_m = [(1, 2), (2, 1), (2, -1), (1, -2),(-1, -2), (-2, -1), (-2, 1), (-1, 2)]

	# checks Knight moves
	for m in knight_m:
		if (s1[0] + m[0], s1[1] + m[1]) == s2:
			return True

	return False


# Q4b

#sqs represent list of squares
def safeConfig(sqs):
    # for each knook
    for i in range(len(sqs)):
        # compare to every other knook
        for ii in range(i+1, len(sqs)):
            # checks if pair is threatened
            if threatening(sqs[i], sqs[ii]):
                return False
    #returns true if there is no pair that is threatened
    return True

 
# Q4c

#helper function for genereating every possible combination of knooks on the board
def allKnooksHelper(sqs):
	allCombs = []

	#iterate through the every square
	for i in range(len(sqs)):
		#create a list of all square other than the current square
		sqsMinus_i = sqs[i] + sqs[i+1]
		#recursively calls the helper to iterate through other square
		allCombsMinus_i = allKnooksHelper(sqs)

		#adds the combination onto the new list
		for combs in allCombsMinus_i:
			allCombs += [combs]

	return allCombs


def allKnooks(n): 
	#generates a list of all squares on the board
	sqs = [(x, y) for x in range(n) for y in range(n)]

	#calls the helper to generate possible combinations
	allCombs = allKnooksHelper(sqs)

	#generates a list of all combinations that are safeConfigs
	validCombs = [combs for combs in allCombs if safeConfig(combs)]

	return validCombs

File: yr2/Semester1/Programming Languagues (F28PL)/CW2/CW2Q4.ml
let rec foldLeft f acc lst =
    match lst with
    | [] -> acc
    | head :: tail -> foldLeft f (f acc head) tail
;;


(* Q4a *)
let threatening (s1, s2) =
	(* list of moves a knight can do *)
	let knightMoves = [(1, 2); (2, 1); (2, -1); (1, -2); (-1, -2); (-2, -1); (-2, 1); (-1, 2)] in	
	
	(* assigns x,y for each point *)
	let (x1, y1) = s1 in
	let (x2, y2) = s2 in
	
	(* checks rook moves by checking for matching x or y *)
	let rookMoveCheck =
		if x1 = x2 || y1 = y2
		then true
		else false 
	in

	(* checks against all variations of knight moves *)
	let rec knightMoveCheck lst =
		match lst with
		|[] -> false
		|(mx,my) :: tl -> 
			if (x1 + mx, y1 + my) = s2
			then true
			else knightMoveCheck tl
	in

	(* returns true if either condition is met *)
	rookMoveCheck || knightMoveCheck knightMoves
;;



(* Q4b *)
let rec safeConfig lst =
	match lst with
	|[] -> true
	|hd1 :: tl -> 
		match tl with
		|[] -> true
		|hd2 :: _ -> 
			(* checks if the 1st and 2nd knooks are threatening *)
			if threatening (hd1, hd2)
			then false
		(* if not, then recursively traverse the list until either a pair is threatened,
			or if it is safe *)	
		else safeConfig tl
;;
